palace/fem/fespace.hpp

```diff
@@ -35,6 +38,167 @@ class FiniteElementSpace : public mfem::ParFiniteElementSpace
+
+public:
+  template <typename... T>
+  AuxiliaryFiniteElementSpace(const FiniteElementSpace &primal_fespace, T &&...args)
+    : FiniteElementSpace(std::forward<T>(args)...), primal_fespace(primal_fespace)
+  {
+  }
+
+  // Return the discrete gradient or discrete curl matrix interpolating from the auxiliary
+  // to the primal space, constructing it on the fly as necessary.
+  const Operator &GetDiscreteInterpolator() const
+  {
+    return G ? *G : BuildDiscreteInterpolator();
+  }
+};
+
+//
+// A collection of FiniteElementSpace objects constructed on the same mesh with the ability
+// to construct the prolongation operators between them as needed.
+//
+class FiniteElementSpaceHierarchy
+{
+protected:
+  std::vector<std::unique_ptr<FiniteElementSpace>> fespaces;
+  mutable std::vector<std::unique_ptr<Operator>> P;
+
+  const Operator &BuildProlongationAtLevel(std::size_t l) const;
+
+public:
+  FiniteElementSpaceHierarchy() = default;
+  FiniteElementSpaceHierarchy(std::unique_ptr<FiniteElementSpace> &&fespace)
+  {
+    AddLevel(std::move(fespace));
+  }
+
+  auto GetNumLevels() const { return fespaces.size(); }
+  auto size() const { return GetNumLevels(); }
+  bool empty() const { return GetNumLevels() == 0; }
+
+  virtual void AddLevel(std::unique_ptr<FiniteElementSpace> &&fespace)
+  {
+    fespaces.push_back(std::move(fespace));
+    P.push_back(nullptr);
+  }
+
+  FiniteElementSpace &GetFESpaceAtLevel(std::size_t l)
+  {
+    MFEM_ASSERT(l >= 0 && l < GetNumLevels(),
+                "Out of bounds request for finite element space at level " << l << "!");
+    return *fespaces[l];
+  }
+  const FiniteElementSpace &GetFESpaceAtLevel(std::size_t l) const
+  {
+    MFEM_ASSERT(l >= 0 && l < GetNumLevels(),
+                "Out of bounds request for finite element space at level " << l << "!");
+    return *fespaces[l];
+  }
+
+  FiniteElementSpace &GetFinestFESpace()
+  {
+    MFEM_ASSERT(!empty(), "Out of bounds request for finite element space at level 0!");
+    return *fespaces.back();
+  }
+  const FiniteElementSpace &GetFinestFESpace() const
+  {
+    MFEM_ASSERT(!empty(), "Out of bounds request for finite element space at level 0!");
+    return *fespaces.back();
+  }
+
+  const Operator &GetProlongationAtLevel(std::size_t l) const
+  {
+    MFEM_ASSERT(l >= 0 && l < GetNumLevels() - 1,
+                "Out of bounds request for finite element space prolongation at level "
+                    << l << "!");
+    return P[l] ? *P[l] : BuildProlongationAtLevel(l);
+  }
+
+  std::vector<const Operator *> GetProlongationOperators() const
+  {
+    std::vector<const Operator *> P_(GetNumLevels() - 1);
+    for (std::size_t l = 0; l < P_.size(); l++)
+    {
+      P_[l] = &GetProlongationAtLevel(l);
+    }
+    return P_;
+  }
+};
+
+//
+// A special type of FiniteElementSpaceHierarchy where all members are auxiliary finite
+// element spaces.
+//
+class AuxiliaryFiniteElementSpaceHierarchy : public FiniteElementSpaceHierarchy
+{
```

Comment:
Not a fan of the slightly tautological interface this ends up with, (ksp.cpp with `aux_fespace->GetFinestAuxiliarySpace()` and then the `AddLevel` override effectively being deleted means this violates the "is a" rule of inheritance. 

I've an alternative suggestion on `hughcars/discrete-interpolator-refactor` that breaks the inheritance of these two spaces whilst reusing the interface, and means the `GetFESpace` methods return auxilliarys directly if needed. Additionally, given the break from mfem already, I'd suggest dropping the `FESpace` from the getter methods, so `GetLevel()` `GetFinestLevel()` etc, as `AddLevel` has already defined what level refers to.

Reply:
I like your approach and have cherry-picked it, in [7df6a19](https://github.com/awslabs/palace/pull/118/commits/7df6a19cd2dfda85a1821beaabc814b79952de3c) with a few minor amendments. Thanks!